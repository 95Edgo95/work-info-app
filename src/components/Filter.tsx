import Button from "components/Button";
import {Link} from "react-router-dom";
import Input from "components/Input";
import * as React from "react";

import * as styles from "assets/css/filter.css";

interface IFilterProps {
  additionalParams?: object;
  attemptGetEntities: (params: object) => void;
  inputs: Array<{
    placeholder: string;
    defaultValue: any;
    name: string;
  }>;
  showAddNew: boolean;
  createLink: string;
}

export default function Filter({ attemptGetEntities, inputs, createLink, additionalParams = {}, showAddNew }: IFilterProps) {
  const setStateMap: any = {};
  const fields: any = {};

  inputs.forEach(({ name, defaultValue }) => {
    const [value, setValue] = React.useState(defaultValue);
    setStateMap[name] = setValue;
    fields[name] = value;
  });

  function handleChange({ currentTarget: { name, value } }) {
    setStateMap[name](value);
  }

  function handleSearch() {
    const searchParams: any = { ...additionalParams };

    Object.keys(fields).forEach((field) => {
      if (fields[field]) {
        searchParams[field] = fields[field]
      }
    });

    attemptGetEntities(searchParams);
  }

  return (
    <div className={styles.filter}>
      {inputs.map(({ name, placeholder }) => (
        <Input
          placeholder={placeholder}
          onChange={handleChange}
          value={fields[name]}
          name={name}
          key={name}
        />
      ))}
      <Button onClick={handleSearch}>
        Search
      </Button>
      {showAddNew && (
        <Button secondary>
          <Link to={`/${createLink}/create`}>
            Add New
          </Link>
        </Button>
      )}
    </div>
  );
}
