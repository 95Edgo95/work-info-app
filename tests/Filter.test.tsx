import {BrowserRouter} from "react-router-dom";
import Filter from "../src/components/Filter";
import { mount } from "enzyme";
import * as React from "react";

const clickFn = jest.fn();

describe('Filter', () => {
  it('should render correctly', () => {
    const component = mount(
      <BrowserRouter>
        <Filter
          attemptGetEntities={clickFn}
          createLink="create"
          inputs={[
            {
              placeholder: "First Name",
              name: "firstName",
              defaultValue: "",
            },
            {
              placeholder: "Last Name",
              defaultValue: "",
              name: "lastName",
            },
            {
              placeholder: "Passport",
              defaultValue: "",
              name: "passport",
            },
            {
              placeholder: "Email",
              defaultValue: "",
              name: "email",
            }
          ]}
        />
      </BrowserRouter>
    );

    expect(component).toMatchSnapshot();
    component.unmount();
  });

  it('should render correctly with additionalParams prop', () => {
    const component = mount(
      <BrowserRouter>
        <Filter
          attemptGetEntities={clickFn}
          createLink="create"
          inputs={[
            {
              placeholder: "First Name",
              name: "firstName",
              defaultValue: "",
            },
            {
              placeholder: "Last Name",
              defaultValue: "",
              name: "lastName",
            },
            {
              placeholder: "Passport",
              defaultValue: "",
              name: "passport",
            },
            {
              placeholder: "Email",
              defaultValue: "",
              name: "email",
            }
          ]}
          additionalParams={{ id: 1}}
        />
      </BrowserRouter>
    );

    expect(component).toMatchSnapshot();
    component.unmount();
  });
});
