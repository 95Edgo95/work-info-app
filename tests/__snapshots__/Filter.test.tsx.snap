// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter should render correctly 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Filter
      attemptGetEntities={[MockFunction]}
      createLink="create"
      inputs={
        Array [
          Object {
            "defaultValue": "",
            "name": "firstName",
            "placeholder": "First Name",
          },
          Object {
            "defaultValue": "",
            "name": "lastName",
            "placeholder": "Last Name",
          },
          Object {
            "defaultValue": "",
            "name": "passport",
            "placeholder": "Passport",
          },
          Object {
            "defaultValue": "",
            "name": "email",
            "placeholder": "Email",
          },
        ]
      }
    >
      <div>
        <Input
          key="firstName"
          name="firstName"
          onChange={[Function]}
          placeholder="First Name"
          value=""
        >
          <div>
            <div
              className=""
            >
              <input
                name="firstName"
                onChange={[Function]}
                placeholder="First Name"
                type="text"
                value=""
              />
            </div>
          </div>
        </Input>
        <Input
          key="lastName"
          name="lastName"
          onChange={[Function]}
          placeholder="Last Name"
          value=""
        >
          <div>
            <div
              className=""
            >
              <input
                name="lastName"
                onChange={[Function]}
                placeholder="Last Name"
                type="text"
                value=""
              />
            </div>
          </div>
        </Input>
        <Input
          key="passport"
          name="passport"
          onChange={[Function]}
          placeholder="Passport"
          value=""
        >
          <div>
            <div
              className=""
            >
              <input
                name="passport"
                onChange={[Function]}
                placeholder="Passport"
                type="text"
                value=""
              />
            </div>
          </div>
        </Input>
        <Input
          key="email"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          value=""
        >
          <div>
            <div
              className=""
            >
              <input
                name="email"
                onChange={[Function]}
                placeholder="Email"
                type="text"
                value=""
              />
            </div>
          </div>
        </Input>
        <Button
          onClick={[Function]}
        >
          <button
            className=""
            onClick={[Function]}
            type="button"
          >
            Search
          </button>
        </Button>
        <Button
          secondary={true}
        >
          <button
            className=""
            type="button"
          >
            <Link
              replace={false}
              to="/create/create"
            >
              <a
                href="/create/create"
                onClick={[Function]}
              >
                Add New
              </a>
            </Link>
          </button>
        </Button>
      </div>
    </Filter>
  </Router>
</BrowserRouter>
`;

exports[`Filter should render correctly with additionalParams prop 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Filter
      additionalParams={
        Object {
          "id": 1,
        }
      }
      attemptGetEntities={[MockFunction]}
      createLink="create"
      inputs={
        Array [
          Object {
            "defaultValue": "",
            "name": "firstName",
            "placeholder": "First Name",
          },
          Object {
            "defaultValue": "",
            "name": "lastName",
            "placeholder": "Last Name",
          },
          Object {
            "defaultValue": "",
            "name": "passport",
            "placeholder": "Passport",
          },
          Object {
            "defaultValue": "",
            "name": "email",
            "placeholder": "Email",
          },
        ]
      }
    >
      <div>
        <Input
          key="firstName"
          name="firstName"
          onChange={[Function]}
          placeholder="First Name"
          value=""
        >
          <div>
            <div
              className=""
            >
              <input
                name="firstName"
                onChange={[Function]}
                placeholder="First Name"
                type="text"
                value=""
              />
            </div>
          </div>
        </Input>
        <Input
          key="lastName"
          name="lastName"
          onChange={[Function]}
          placeholder="Last Name"
          value=""
        >
          <div>
            <div
              className=""
            >
              <input
                name="lastName"
                onChange={[Function]}
                placeholder="Last Name"
                type="text"
                value=""
              />
            </div>
          </div>
        </Input>
        <Input
          key="passport"
          name="passport"
          onChange={[Function]}
          placeholder="Passport"
          value=""
        >
          <div>
            <div
              className=""
            >
              <input
                name="passport"
                onChange={[Function]}
                placeholder="Passport"
                type="text"
                value=""
              />
            </div>
          </div>
        </Input>
        <Input
          key="email"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          value=""
        >
          <div>
            <div
              className=""
            >
              <input
                name="email"
                onChange={[Function]}
                placeholder="Email"
                type="text"
                value=""
              />
            </div>
          </div>
        </Input>
        <Button
          onClick={[Function]}
        >
          <button
            className=""
            onClick={[Function]}
            type="button"
          >
            Search
          </button>
        </Button>
        <Button
          secondary={true}
        >
          <button
            className=""
            type="button"
          >
            <Link
              replace={false}
              to="/create/create"
            >
              <a
                href="/create/create"
                onClick={[Function]}
              >
                Add New
              </a>
            </Link>
          </button>
        </Button>
      </div>
    </Filter>
  </Router>
</BrowserRouter>
`;
